mi/dataset/parser/pco2a_a_sample.py

```diff
@@ -0,0 +1,268 @@
+    ('measured_water_co2', SENSOR_GROUP_CO2, float),
+    ('avg_irga_temperature', SENSOR_GROUP_AVG_IRGA_TEMP, float),
+    ('humidity', SENSOR_GROUP_HUMIDITY, float),
+    ('humidity_temperature', SENSOR_GROUP_HUMIDITY_TEMP, float),
+    ('gas_stream_pressure', SENSOR_GROUP_STREAM_PRESSURE, int),
+    ('irga_detector_temperature', SENSOR_GROUP_DETECTOR_TEMP, float),
+    ('irga_source_temperature', SENSOR_GROUP_SOURCE_TEMP, float),
+    ('supply_voltage', SENSOR_GROUP_SUPPLY_VOLTAGE, float)
+]
+
+
+class DataParticleType(BaseEnum):
+    PCO2A_INSTRUMENT_AIR_PARTICLE = 'pco2a_a_dcl_instrument_air'
+    PCO2A_INSTRUMENT_WATER_PARTICLE = 'pco2a_a_dcl_instrument_water'
+#   PCO2A_INSTRUMENT_AIR_RECOVERED_PARTICLE = 'pco2a_a_dcl_instrument_air_recovered'
+#   PCO2A_INSTRUMENT_WATER_RECOVERED_PARTICLE = 'pco2a_a_dcl_instrument_water_recovered'
+
+
+class Pco2aADclParticleClassKey(BaseEnum):
+    """
+    An enum for the keys application to the pco2a_a_dcl particle classes
+    """
+    AIR_PARTICLE_CLASS = 'air_particle_class'
+    WATER_PARTICLE_CLASS = 'water_particle_class'
+
+
+class Pco2aADclInstrumentDataParticleAir(DclInstrumentDataParticle):
+    """
+    Class for generating the Pco2a_a_dcl instrument particles.
+    """
+    data_matcher = SENSOR_DATA_MATCHER_AIR
+
+    def __init__(self, raw_data, *args, **kwargs):
+        super(Pco2aADclInstrumentDataParticleAir, self).__init__(
+            raw_data, INSTRUMENT_PARTICLE_AIR_MAP, *args, **kwargs)
+
+        # instrument_timestamp is the internal_timestamp
+        instrument_timestamp = self.raw_data[SENSOR_GROUP_SENSOR_DATE_TIME]
+        elapsed_seconds_useconds = \
+            timestamp_yyyy_mm_dd_hh_mm_ss_csv_to_ntp(instrument_timestamp)
+        self.set_internal_timestamp(elapsed_seconds_useconds)
+
+        # instrument clock is not accurate so, use port_timestamp as the preferred_ts
+        self.contents[DataParticleKey.PREFERRED_TIMESTAMP] = DataParticleKey.PORT_TIMESTAMP
+
+
+class Pco2aADclInstrumentDataParticleWater(DclInstrumentDataParticle):
+    """
+    Class for generating the Pco2a_a_dcl instrument particles.
+    """
+    data_matcher = SENSOR_DATA_MATCHER_WATER
+
+    def __init__(self, raw_data, *args, **kwargs):
+        super(Pco2aADclInstrumentDataParticleWater, self).__init__(
+            raw_data, INSTRUMENT_PARTICLE_WATER_MAP, *args, **kwargs)
+
+        # Instrument timestamp is the internal timestamp
+        instrument_timestamp = self.raw_data[SENSOR_GROUP_SENSOR_DATE_TIME]
+        elapsed_seconds_useconds = \
+            timestamp_yyyy_mm_dd_hh_mm_ss_csv_to_ntp(instrument_timestamp)
+        self.set_internal_timestamp(elapsed_seconds_useconds)
+
+        # instrument clock is not accurate so, use port_timestamp as the preferred_ts
+        self.contents[DataParticleKey.PREFERRED_TIMESTAMP] = DataParticleKey.PORT_TIMESTAMP
+
+
+class Pco2aADclTelemeteredInstrumentDataParticleAir(Pco2aADclInstrumentDataParticleAir):
```

Comment:
I think you only need `Pco2aADclInstrumentDataParticleAir` and then point the references of both the telemetered and recovered driver to the one class.

Reply:
Done